libsolidity/experimental/ast/TypeSystemHelper.cpp

```diff
@@ -287,9 +287,38 @@ std::vector<experimental::Type> TypeEnvironmentHelpers::typeVars(Type _type) con
 	};
 	typeVarsImpl(_type, typeVarsImpl);
 	return typeVars;
-
 }
 
+experimental::Type TypeEnvironmentHelpers::substitute(
+	Type const& _type,
+	TypeVariable const& _variableToReplace,
```

Comment:
So the implementation should just do the following:
If ``_type`` is equal to ``_toReplace`` return ``_replacement``.
If not and it's a type constant: recurse on the arguments of the type constant.
Otherwise: return ``_type``.

But yeah, actually not sure if we need this anymore in the first place :-).

Reply:
ok. I changed it to work on any `Type` they way you described it.

> But yeah, actually not sure if we need substitution anymore in the first place :-).

I thought you were fine with it either way last time we discussed it. I can change it to unification in a cloned env, but then I still need to introduce some new element:
- Some way to freshen just selected variables within a type. Or
- Some way to unmark selected variables as fixed. `generalizeTypeVars()` could be used for that but you said we'd never do that. Or
- Some way to clone while unmarking selected variables as fixed. But this is just `clone()` and `generalizeTypeVars()` combined.

The first one can be done with `substitute()` so it's what I went for. Should I switch to one of the others?

Reply:
There is only one type variable in the types in question, isn't it? So you don't need to freshen just selected variables.
You also don't need to unmark variables from fixed, since freshening will create, well, fresh ones.
You also don't need a new type environment.
It's literally freshening, fetching the one type variable of the result, and unifying that with the substituent, isn't it?

Reply:
But I'm also fine with keeping it, if it's rewritten properly. The current rewrite looks wrong to me, though. You need to check for type equality *before* descending into type constants.

Reply:
> There is only one type variable in the types in question, isn't there? So you don't need to freshen just selected variables.

Ah, you're right. With type class functions that's just one variable.


> You need to check for type equality *before* descending into type constants.

Right. Fixed. I was too quick to push this, assuming my tests would have caught such problems :)